internal/packages/walk.go

```diff
@@ -421,8 +428,30 @@ func excludedForSubdir(excluded []string, subdir string) []string {
 }
 
 type symlinkResolver struct {
-	root    string
-	visited []string
+	root string
+
+	// visited maps canonical absolute directory paths (not containing symlink
+	// components) to the paths where they were actually visited.
+	visited map[string]string
+}
+
+func newSymlinkResolver(root string) *symlinkResolver {
+	return &symlinkResolver{root: root, visited: make(map[string]string)}
+}
+
+// visit finds the canonical path for a directory and marks it as visited. If
+// the directory was already visited (via a symlink for example), visit returns
+// an error.
+func (r *symlinkResolver) visit(dir string) error {
```

Comment:
I split `follow` into two methods: `visit` is called for every directory we visit. It returns an error if we visit the same directory (checked using canonical path). `follow` checks whether we should recurse on a symlink.